lib/AtomicCache.php

```diff
@@ -0,0 +1,362 @@
+<?php
+
+namespace Amp\Cache;
+
+use Amp\Promise;
+use Amp\Serialization\SerializationException;
+use Amp\Sync\KeyedMutex;
+use Amp\Sync\Lock;
+use function Amp\call;
+
+final class AtomicCache
+{
+    /** @var SerializedCache */
+    private $cache;
+
+    /** @var KeyedMutex */
+    private $mutex;
+
+    /**
+     * @param SerializedCache $cache
+     * @param KeyedMutex      $mutex
+     */
+    public function __construct(SerializedCache $cache, KeyedMutex $mutex)
+    {
+        $this->cache = $cache;
+        $this->mutex = $mutex;
+    }
+
+    /**
+     * Obtains the lock for the given key, then invokes the $create callback with the current cached value (which may
+     * be null if the key did not exist in the cache). The value returned from the callback is stored in the cache and
+     * the promise returned from this method is resolved with the value.
+     *
+     * @param string   $key
+     * @param callable(string $key, mixed|null $value): mixed $create
+     * @param int|null $ttl Timeout in seconds. The default `null` $ttl value indicates no timeout.
+     *
+     * @return Promise<mixed>
+     *
+     * @throws CacheException If the $create callback throws an exception while generating the value.
+     * @throws SerializationException If serializing the value returned from the callback fails.
+     */
+    public function compute(string $key, callable $create, ?int $ttl = null): Promise
+    {
+        return call(function () use ($key, $create, $ttl): \Generator {
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            try {
+                $value = yield $this->cache->get($key);
+
+                return yield from $this->create($create, $key, $value, $ttl);
+            } finally {
+                $lock->release();
+            }
+        });
+    }
+
+    /**
+     * Attempts to get the value for the given key. If the key is not found, the key is locked, the $create callback
+     * is invoked with the key as the first parameter and null as the second parameter. The value returned from the
+     * callback is stored in the cache and the promise returned from this method is resolved with the value.
+     *
+     * @param string   $key Cache key.
+     * @param callable(string $key, null $value): mixed $create
```

Comment:
Does it make sense to even provide the second parameter here?

Reply:
It was easier to have a single method to invoke the function even when the value was null. Should I remove it from the docs though?